internal/testtypes/builders.go

```diff
@@ -0,0 +1,562 @@
+func (m *MachineDeploymentBuilder) WithBootstrapTemplate(ref *unstructured.Unstructured) *MachineDeploymentBuilder {
+	m.bootstrapTemplate = ref
+	return m
+}
+
+// WithInfrastructureTemplate adds the passed unstructured object to the MachineDeployment builder as an infrastructureMachineTemplate.
+func (m *MachineDeploymentBuilder) WithInfrastructureTemplate(ref *unstructured.Unstructured) *MachineDeploymentBuilder {
+	m.infrastructureTemplate = ref
+	return m
+}
+
+// WithLabels adds the given labels to the MachineDeploymentBuilder.
+func (m *MachineDeploymentBuilder) WithLabels(labels map[string]string) *MachineDeploymentBuilder {
+	m.labels = labels
+	return m
+}
+
+// WithReplicas sets the number of replicas for the MachineDeploymentClassBuilder.
+func (m *MachineDeploymentBuilder) WithReplicas(replicas *int32) *MachineDeploymentBuilder {
+	m.replicas = replicas
+	return m
+}
+
+// Build creates a new MachineDeployment with the variables and objects passed to the MachineDeploymentBuilder.
+func (m *MachineDeploymentBuilder) Build() *clusterv1.MachineDeployment {
+	obj := &clusterv1.MachineDeployment{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "MachineDeployment",
+			APIVersion: clusterv1.GroupVersion.String(),
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      m.name,
+			Namespace: m.namespace,
+			Labels:    m.labels,
+		},
+	}
+	obj.Spec.Replicas = m.replicas
+	if m.bootstrapTemplate != nil {
+		obj.Spec.Template.Spec.Bootstrap.ConfigRef = objToRef(m.bootstrapTemplate)
+	}
+	if m.infrastructureTemplate != nil {
+		obj.Spec.Template.Spec.InfrastructureRef = *objToRef(m.infrastructureTemplate)
+	}
+	return obj
+}
+
+// objToRef returns a reference to the given object.
+func objToRef(obj client.Object) *corev1.ObjectReference {
+	gvk := obj.GetObjectKind().GroupVersionKind()
+	return &corev1.ObjectReference{
+		Kind:       gvk.Kind,
+		APIVersion: gvk.GroupVersion().String(),
+		Namespace:  obj.GetNamespace(),
+		Name:       obj.GetName(),
+	}
+}
+
+// setNestedRef sets the value of a nested field to a reference to the refObj provided.
+func setNestedRef(obj, refObj *unstructured.Unstructured, fields ...string) error {
+	ref := map[string]interface{}{
+		"kind":       refObj.GetKind(),
+		"namespace":  refObj.GetNamespace(),
+		"name":       refObj.GetName(),
+		"apiVersion": refObj.GetAPIVersion(),
+	}
+	return unstructured.SetNestedField(obj.UnstructuredContent(), ref, fields...)
+}
+
+// setSpecFields sets fields in an unstructured object from a map.
+func setSpecFields(obj *unstructured.Unstructured, fields map[string]interface{}) {
+	for k, v := range fields {
+		fieldParts := strings.Split(k, ".")
+		if len(fieldParts) == 0 {
+			panic(fmt.Errorf("fieldParts invalid"))
+		}
+		if fieldParts[0] != "spec" {
+			panic(fmt.Errorf("can not set fields outside spec"))
+		}
+		// TODO(killianmuldoon): Update to use the internal/contract package
```

Comment:
```suggestion
```
@killianmuldoon sorry last one then lgtm :). I don't think we need this TODO
(ref: https://github.com/kubernetes-sigs/cluster-api/pull/5055#discussion_r695668540)

I think you can also squash already.